pkg/cmd/cli/cmd/edit.go

```diff
@@ -0,0 +1,333 @@
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/api/latest"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/editor"
+)
+
+const (
+	edit_long = `Edit a resource from the default editor.
+
+The edit command allows you to directly edit any API resource you can retrieve via the
+command line tools. It will open the editor defined by your OSC_EDITOR, GIT_EDITOR,
+or EDITOR environment variables, or fall back to 'vi'. You can edit multiple objects,
+although changes are applied one at a time. The command accepts filenames as well as
+command line arguments, although the files you point to must be previously saved
+versions of resources.
+
+The files to edit will be output in the default API version, or a version specified
+by --output-version. The default format is YAML - if you would like to edit in JSON
+pass -o json.
+
+In the event an error occurs while updating, a temporary file will be created on disk
+that contains your unapplied changes. The most common error when updating a resource
+is another editor changing the resource on the server. When this occurs, you will have
+to apply your changes to the newer version of the resource, or update your temporary
+saved copy to include the latest resource version.
+
+Examples:
+
+	# Edit the service named 'docker-registry':
+	$ %[1]s edit svc/docker-registry
+
+	# Edit the deployment config named 'my-deployment':
+	$ %[1]s edit dc/my-deployment
+
+	# Use an alternative editor
+	$ OSC_EDITOR="nano" %[1]s edit dc/my-deployment
+
+	# Edit the service 'docker-registry' in JSON using the v1beta3 API format:
+	$ %[1]s edit svc/docker-registry --output-version=v1beta3 -o json
+`
+)
+
+func NewCmdEdit(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	var filenames util.StringList
+	cmd := &cobra.Command{
+		Use:   "edit -f FILENAME",
+		Short: "Edit a resource on the server and apply the update.",
+		Long:  fmt.Sprintf(edit_long, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunEdit(fullName, f, out, cmd, args, filenames)
+			if err == errExit {
+				os.Exit(1)
+			}
+			cmdutil.CheckErr(err)
+		},
+	}
+	cmd.Flags().StringP("output", "o", "yaml", "Output format. One of: yaml|json.")
+	cmd.Flags().String("output-version", "", "Output the formatted object with the given version (default api-version).")
+	cmd.Flags().VarP(&filenames, "filename", "f", "Filename, directory, or URL to file to use to edit the resource.")
+	return cmd
+}
+
+func RunEdit(fullName string, f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string, filenames util.StringList) error {
+	var p kubectl.ResourcePrinter
```

Comment:
I think I'm going to fix the versioning problem first.

----- Original Message -----
> > +		Run: func(cmd *cobra.Command, args []string) {
> > +			err := RunEdit(fullName, f, out, cmd, args, filenames)
> > +			if err == errExit {
> > +				os.Exit(1)
> > +			}
> > +			cmdutil.CheckErr(err)
> > +		},
> > +	}
> > +	cmd.Flags().StringP("output", "o", "yaml", "Output format. One of:
> > yaml|json.")
> > +	cmd.Flags().String("output-version", "", "Output the formatted object
> > with the given version (default api-version).")
> > +	cmd.Flags().VarP(&filenames, "filename", "f", "Filename, directory, or
> > URL to file to use to edit the resource.")
> > +	return cmd
> > +}
> > +
> > +func RunEdit(fullName string, f *clientcmd.Factory, out io.Writer, cmd
> > *cobra.Command, args []string, filenames util.StringList) error {
> > +	var p kubectl.ResourcePrinter
> 
> `unversionedPrinter` instead of `p`?  The usage being so far from the
> declaration makes it hard to remember.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1758/files#r28512058